### PR TITLE
S-69406 mora label placement improvement

### PR DESCRIPTION
**Pull request by JD31**

### CODE REVIEW

File: src/osgEarth/MPScreenSpaceLayoutSG.cpp [hunk 1]
```diff
@@ -453,13 +464,14 @@ struct /*internal*/ MPDeclutterSortSG : public osgUtil::RenderBin::SortCallback
...
             if ( ! leaf->_drawable.valid() )
+            {
+                printf("\nleaf->_drawable.valid() is not valid");
                 continue;
-
+            }
             MPScreenSpaceGeometry* annoDrawable = static_cast<MPScreenSpaceGeometry*>(leaf->_drawable.get());
 
             // transform the bounding box of the drawable into window-space.
-            //osg::BoundingBox box = annoDrawable->isAutoFollowLine() ? annoDrawable->getBBox() : annoDrawable->getBoundingBox();
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG.cpp]: Why did you uncomment ?

File: src/osgEarth/MPScreenSpaceLayoutSG.cpp [hunk 2]
```diff
@@ -453,13 +464,14 @@ struct /*internal*/ MPDeclutterSortSG : public osgUtil::RenderBin::SortCallback
         {
             osgUtil::RenderLeaf* leaf = *i;
             if ( ! leaf->_drawable.valid() )
+            {
+                printf("\nleaf->_drawable.valid() is not valid");
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG.cpp]: Remove the trace

File: src/osgEarth/MPScreenSpaceLayoutSG [hunk 3]
```diff
@@ -128,11 +146,12 @@ public:
     long       _id{1};
     double     _originalPriority{0.};
     double     _priority{0.};
+    double     _textPolygonAltitude;
     bool       _originalDeclutterActivated{true};
     bool       _declutterActivated{true};
     osg::Vec2s _pixelOffset{0, 0};
     bool       _autoRotate{false}, _autoFollowLine{false};
-    osg::Vec3d _anchorPoint, _lineStart, _lineEnd;
+    osg::Vec3d _anchorPoint, _lineStart, _lineEnd,_poly1, _poly2;
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG]: space before _poly1

File: src/osgEarth/MPScreenSpaceLayoutSG.cpp [hunk 4]
```diff
@@ -482,14 +494,80 @@ struct /*internal*/ MPDeclutterSortSG : public osgUtil::RenderBin::SortCallback
...
                 annoDrawable->_cull_anchorOnScreen += slidingOffset;
             }
-            
+
+            // Computes label location for the grid mora (visible part of the polygon)
+            if (annoDrawable->polygonVisible())
+            {
+                double maxAlt = annoDrawable->getTextPolygonAltitude();
+                double currentAltitude;
+                cam->getUserValue("altitude", currentAltitude);
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG.cpp]: Not here !
It is camera related and is then const for the whole frame.
(should be moved where the eye, look,... are computed)

File: src/osgEarth/MPScreenSpaceLayoutSG.cpp [hunk 5]
```diff
@@ -482,14 +494,80 @@ struct /*internal*/ MPDeclutterSortSG : public osgUtil::RenderBin::SortCallback
...
+                    else
+                    {
+                        osg::Vec3d pc[5];
+                        pc[0] = annoDrawable->getLineStartPoint()* MVP;
+                        pc[2] = annoDrawable->getLineEndPoint()* MVP;
+                        pc[1] = annoDrawable->getPolygonPoint1()* MVP;
+                        pc[3] = annoDrawable->getPolygonPoint2()* MVP;
+                        pc[4] = pc[0];      // to close the polygon (repeat 1st point)
+
+                        boost_polygon geomMora;
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG.cpp]: mora is a specific case (you can say it if you want in the comments) but don't name a variable *mora*

File: src/osgEarth/MPScreenSpaceLayoutSG.cpp [hunk 6]
```diff
@@ -482,14 +494,80 @@ struct /*internal*/ MPDeclutterSortSG : public osgUtil::RenderBin::SortCallback
...
+
+
             // computes the clamped labels (used for graticules)
             if (annoDrawable->screenClamping())
             {
                 const osgEarth::SpatialReference* srs = osgEarth::SpatialReference::create("epsg:4326");
                 
                 // Calculate the "clip to world" matrix = MVPinv.
-                osg::Matrix MVP = cam->getViewMatrix() * cam->getProjectionMatrix();
+//                osg::Matrix MVP = cam->getViewMatrix() * cam->getProjectionMatrix();
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG.cpp]: Can be entirely removed I guess ?

File: src/osgEarth/MPScreenSpaceLayoutSG.cpp [hunk 7]
```diff
@@ -660,12 +739,13 @@ struct /*internal*/ MPDeclutterSortSG : public osgUtil::RenderBin::SortCallback
                 }
 
                 // A max priority => always display
-                if ( annoDrawable->_priority == DBL_MAX || ! annoDrawable->_declutterActivated)
-                {
-                    visible = true;
-                }
+//                if ( annoDrawable->_priority == DBL_MAX || ! annoDrawable->_declutterActivated)
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG.cpp]: Don't comment code. Just remove

**Review by MPcoreDev — COMMENTED** [review]
Quality review

File: src/osgEarth/MPScreenSpaceLayoutSG [hunk 8]
```diff
@@ -56,16 +56,22 @@ public:
...
+                              _instanceIndex(0),
                               _screenClamping(false),
-                              _instanceIndex(0) { }
+                              _polygonVisible(false){ }
 
     /** Decluttering priority - FLT_MAX means don't declutter */
     void setPriority(double value) { _originalPriority = _priority = value; }
     double getPriority() const     { return _priority; }
 
+    /** Decluttering priority - FLT_MAX means don't declutter */
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG]: Bad comment

File: src/osgEarth/MPScreenSpaceLayoutSG [hunk 9]
```diff
@@ -56,16 +56,22 @@ public:
                               _id(-1),
                               _originalPriority(0.0),
                               _priority(0.0),
+                              _textPolygonAltitude(0.0),
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG]: Should be _MAX by default as if not defined, but "polygonVisible" defined, we expect the algo to run at all altitude by default

File: src/osgEarth/MPScreenSpaceLayoutSG [hunk 10]
```diff
@@ -106,6 +120,10 @@ public:
     void setScreenClamping(bool screenClamping) { _screenClamping = screenClamping; }
     bool screenClamping() const { return _screenClamping; }
 
+    /** Whether the label shall be displayed on the visible part of the polygon */
+    void setPolygonVisible(bool polygonVisible) {_polygonVisible = polygonVisible; }
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG]: space before _

File: src/osgEarthAnnotation/MPAnnotationGroupSG.cpp [hunk 11]
```diff
@@ -357,6 +356,54 @@ long MPAnnotationGroupSG::addAnnotation(const Style& style, Geometry *geom, cons
...
+        else
+        {
+            OE_WARN << "no geomPolygon avail" << std::endl;
+        }
+        annoDrawable->setLineStartPoint(p1);
+        annoDrawable->setLineEndPoint(p2);
+        annoDrawable->setPolygonPoint1(p3);
+        annoDrawable->setPolygonPoint2(p4);
+        annoDrawable->setPolygonVisible(true);
+        annoDrawable->setTextPolygonAltitude(ts->textPolygonAltitude().get());
```
Comment (MPcoreDev) [inline src/osgEarthAnnotation/MPAnnotationGroupSG.cpp]: textPolygonAltitude is optional.
what about when it is not defined ?

**Review by MPcoreDev — COMMENTED** [review]
Review Number 2

File: src/osgEarth/MPScreenSpaceLayoutSG (hunk 9 above)
Comment (ggranadosi) [reply src/osgEarth/MPScreenSpaceLayoutSG]: actuallly, it is the opposite. CurrentAltitude is from the camera and if maxAlt is equal to 0, it won't use the algorithm.
            
            if (annoDrawable->polygonVisible())
            {
                double maxAlt = annoDrawable->getTextPolygonAltitude();
                if (currentAltitude > maxAlt)
                {
                    visible = true;
                }
                else
                {
                    visible = false;

Comment (ggranadosi) [reply src/osgEarth/MPScreenSpaceLayoutSG]: updated as per your suggestion.

File: src/osgEarth/MPScreenSpaceLayoutSG [hunk 12]
```diff
@@ -128,11 +146,12 @@ public:
     long       _id{1};
     double     _originalPriority{0.};
     double     _priority{0.};
+    double     _textPolygonAltitude;
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG]: Please add initialization value like for other attributes.

File: src/osgEarth/MPScreenSpaceLayoutSG.cpp [hunk 13]
```diff
@@ -442,6 +450,11 @@ struct /*internal*/ MPDeclutterSortSG : public osgUtil::RenderBin::SortCallback
         // has the camera moved?
         //        bool camChanged = camVPW != local._lastCamVPW;
         local._lastCamVPW = camVPW;
+        osg::Matrix MVP = cam->getViewMatrix() * cam->getProjectionMatrix();
+        typedef boost::geometry::model::d2::point_xy<double> boost_point;
+        typedef boost::geometry::model::polygon<boost_point> boost_polygon;
+        boost_polygon geomScreen;
+        boost::geometry::read_wkt("POLYGON((-1.0 -1.0 , -1.0 1.0 , 1.0 1.0 , 1.0 -1.0, -1.0 -1.0))", geomScreen);
```
Comment (MPcoreDev) [inline src/osgEarth/MPScreenSpaceLayoutSG.cpp]: Don't do that at each frame. But only once for the program.

**Review by MPcoreDev — CHANGES_REQUESTED** [review]
Quality review 3

**Pull request merged**